### PR TITLE
MRG, ENH: Fix Report bug and add to CircleCI

### PR DESCRIPTION
**Pull request by larsoner**
Closes #5894.

Also adds the Report HTML to CircleCI build + output using a custom scraper.

WIP only because it requires https://github.com/sphinx-gallery/sphinx-gallery/pull/502 to be merged, then we can revert the `.circleci/config.yaml` change where I make it use by branch.

@massich @jasmainak can you see if this fixes your segfault problem? Thanks @jasmainak for the GUI reminder, it ended up fixing the problem for me at least.

### CODE REVIEW

**Comment by larsoner** [comment]
Almost works. It doesn't render quite correctly when built by CircleCI:

https://13624-1301584-gh.circle-artifacts.com/0/dev/auto_examples/visualization/plot_make_report.html

It doesn't do this locally so I'll have to try to figure out what the problem is. I'll probably split the `Report` fix into a separate PR so that can at least land in `master`.

**Comment by larsoner** [comment]
Renders somewhat oddly on my macOS machine on Chrome and Safari. But even the report HTML file itself renders the same way, so I think the Sphinx-gallery bits are working now at least:

https://13634-1301584-gh.circle-artifacts.com/0/dev/auto_examples/visualization/plot_make_report.html

**Comment by agramfort** [comment]
there is a weird interaction with the javascript of the report (buttons
etc.)
is the notebook in an iframe?

>

**Comment by larsoner** [comment]
> is the notebook in an iframe?

Yes with `<iframe src="...">`. The first thing I tried was `<iframe srcdoc="...">` where you actually insert the HTML in the `srcdoc` field. But with this the javascript was even worse (clicking did nothing). Not sure what other options there are unfortunately. We could just not embed it and give just the link to the rendered HTML.

**Comment by agramfort** [comment]
>
> > We could just not embed it and give just the link to the rendered HTML.
>
+10

**Comment by jasmainak** [comment]
I was trying to use the HTML from the report (which is stored as a list of strings) to display it in a jupyter notebook but had only limited success. Linking is probably the easiest option.

File: mne/report.py [hunk 1]
```diff
@@ -2058,3 +2062,60 @@ def _fix_global_ids(html):
...
+_SCRAPER_TEXT = '''
+.. only:: builder_html
+
+    .. container:: row
+
+        .. rubric:: The `HTML document <{0}>`__ written by :meth:`mne.Report.save`:
+
+        .. raw:: html
+
+            <iframe class="sg_report" sandbox="allow-scripts" src="{0}"></iframe>
```
Comment (jasmainak) [inline mne/report.py]: what if you just put the entire report string here? ... just a thought.

Comment (larsoner) [reply mne/report.py]: If you mean ..raw:html directly, the report has html header information and tags so it should work worse. If you mean srcdoc in the inline tag I tried and it was worse than a separate file

Comment (jasmainak) [reply mne/report.py]: okay let's keep it separate

**Comment by larsoner** [comment]
Okay I figured out that the rendering problems I was seeing with the embedding were due to the SVG images on CircleCI, probably some corner case involving `dpi`. I switched the example to use PNG and now the rendering and interaction works as expected for me, at least on Linux+Chrome:

https://13662-1301584-gh.circle-artifacts.com/0/dev/auto_examples/visualization/plot_make_report.html

@jasmainak @agramfort can you try again?

**Comment by agramfort** [comment]
works great ! thx @larsoner

**Comment by jasmainak** [comment]
sorry for slow responses. I don't know how @agramfort manages to get internet in the conference center.

But wow, it looks great! You might be able to motivate me to write a good tutorial on report now that we have this :)

File: doc/conf.py [hunk 2]
```diff
@@ -316,19 +316,27 @@
...
+    report_scraper = None
 else:
     # Let's do the same thing we do in tests: reraise traits exceptions
     from traits.api import push_exception_handler
     push_exception_handler(reraise_exceptions=True)
+    report_scraper = mne.report._ReportScraper()
+    scrapers = ('matplotlib', 'mayavi', report_scraper)
+
+
+def setup(app):
```
Comment (jasmainak) [inline doc/conf.py]: out of curiosity, where is this variable `app` defined?

and why don't you have to provide the `scrapers` variable to sphinx gallery config?

**Comment by massich** [comment]
yep its great !!

**Comment by massich** [comment]
Do you guys think we should backport this one? It would simplify https://github.com/mne-tools/mne-study-template/pull/4

File: doc/conf.py (hunk 2 above)
Comment (larsoner) [reply doc/conf.py]: `app` is provided by Sphinx when it calls this function

`scrapers` is modified above in the `mayavi` try/except clause

**Comment by larsoner** [comment]
> You might be able to motivate me to write a good tutorial on report now that we have this :)

We actually already have this example and `doc/tutorials/report.rst`. They should be merged. All of the content is probably there and it should mostly need to be merged. Can you take a look?

> Do you guys think we should backport this one?

Yes but let's backport just the Report fix commit. I'll open a PR

**Pull request merged**